File: app/src/test/kotlin/com/deepseek/harness/android/ui/chat/TimelineGroupingTest.kt
package com.deepseek.harness.android.ui.chat

import com.deepseek.harness.android.domain.model.ChatMessage
import com.deepseek.harness.android.domain.model.MessageRole
import com.deepseek.harness.android.domain.model.TimelineItem
import org.junit.Assert.assertEquals
import org.junit.Test
class TimelineGroupingTest {

    private fun message(id: String): TimelineItem.Message =
        TimelineItem.Message(
            ChatMessage(id = id, sessionId = "s1", role = MessageRole.USER, text = id),
        )

    @Test
    fun `items before the first boundary form the null group`() {
        val groups = groupTimelineByTurn(listOf(message("a"), message("b")))

        assertEquals(1, groups.size)
        assertEquals(null, groups.single().turn)
        assertEquals(listOf("a", "b"), groups.single().items.map { (it as TimelineItem.Message).value.id })
    }

    @Test
    fun `boundaries split groups and land only in the heading`() {
        val groups = groupTimelineByTurn(
            listOf(
                TimelineItem.TurnBoundary(1L),
                message("a"),
                TimelineItem.TurnBoundary(2L),
                message("b"),
                message("c"),
            ),
        )

        assertEquals(3, groups.size)
        assertEquals(null, groups[0].turn)
        assertEquals(0, groups[0].items.size)
        assertEquals(1L, groups[1].turn)
        assertEquals(listOf("a"), groups[1].items.map { (it as TimelineItem.Message).value.id })
        assertEquals(2L, groups[2].turn)
        assertEquals(listOf("b", "c"), groups[2].items.map { (it as TimelineItem.Message).value.id })
    }

    @Test
    fun `a boundary without rows keeps its group for the outline number`() {
        val groups = groupTimelineByTurn(
            listOf(
                TimelineItem.TurnBoundary(1L),
                message("a"),
                TimelineItem.TurnBoundary(2L),
            ),
        )

        assertEquals(3, groups.size)
        assertEquals(2L, groups[2].turn)
        assertEquals(0, groups[2].items.size)
    }

    @Test
    fun `prompt preview echoes the first user message folded to one line`() {
        val preview = promptPreview(
            listOf(
                messageWith(id = "u1", role = MessageRole.ASSISTANT, text = "prior answer"),
                messageWith(id = "u2", role = MessageRole.USER, text = "first line\nsecond line"),
                messageWith(id = "u3", role = MessageRole.USER, text = "later prompt"),
            ),
        )

        assertEquals("first line", preview)
    }

    @Test
    fun `prompt preview truncates long prompts and returns null without user rows`() {
        val long = promptPreview(
            listOf(messageWith(id = "u1", role = MessageRole.USER, text = "x".repeat(100))),
        )
        val none = promptPreview(
            listOf(messageWith(id = "a1", role = MessageRole.ASSISTANT, text = "answer")),
        )

        assertEquals(60, long?.length)
        assertEquals("…", long?.takeLast(1))
        assertEquals(null, none)
    }

    private fun messageWith(id: String, role: MessageRole, text: String): TimelineItem.Message =
        TimelineItem.Message(
            ChatMessage(id = id, sessionId = "s1", role = role, text = text),
        )
}

File: flutter/app/test/ui/chat/timeline_grouping_test.dart
import 'package:domain/model/chat_message.dart';
import 'package:domain/model/timeline_item.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:app/ui/chat/timeline_grouping.dart';

TimelineMessage message(String id) => TimelineMessage(
      ChatMessage(id: id, sessionId: 's1', role: MessageRole.user, text: id),
    );

TimelineMessage messageWith({
  required String id,
  required MessageRole role,
  required String text,
}) =>
    TimelineMessage(
      ChatMessage(id: id, sessionId: 's1', role: role, text: text),
    );

void main() {
  test('items before the first boundary form the null group', () {
    final groups = groupTimelineByTurn(<TimelineItem>[message('a'), message('b')]);

    expect(groups, hasLength(1));
    expect(groups.single.turn, isNull);
    expect(
      groups.single.items
          .whereType<TimelineMessage>()
          .map((item) => item.value.id)
          .toList(),
      <String>['a', 'b'],
    );
  });

  test('boundaries split groups and land only in the heading', () {
    final groups = groupTimelineByTurn(<TimelineItem>[
      const TimelineTurnBoundary(1),
      message('a'),
      const TimelineTurnBoundary(2),
      message('b'),
      message('c'),
    ]);

    expect(groups, hasLength(3));
    expect(groups[0].turn, isNull);
    expect(groups[0].items, isEmpty);
    expect(groups[1].turn, 1);
    expect(
      groups[1].items
          .whereType<TimelineMessage>()
          .map((item) => item.value.id)
          .toList(),
      <String>['a'],
    );
    expect(groups[2].turn, 2);
    expect(
      groups[2].items
          .whereType<TimelineMessage>()
          .map((item) => item.value.id)
          .toList(),
      <String>['b', 'c'],
    );
  });

  test('a boundary without rows keeps its group for the outline number', () {
    final groups = groupTimelineByTurn(<TimelineItem>[
      const TimelineTurnBoundary(1),
      message('a'),
      const TimelineTurnBoundary(2),
    ]);

    expect(groups, hasLength(3));
    expect(groups[2].turn, 2);
    expect(groups[2].items, isEmpty);
  });

  test('prompt preview echoes the first user message folded to one line', () {
    final preview = promptPreview(<TimelineItem>[
      messageWith(id: 'u1', role: MessageRole.assistant, text: 'prior answer'),
      messageWith(id: 'u2', role: MessageRole.user, text: 'first line\nsecond line'),
      messageWith(id: 'u3', role: MessageRole.user, text: 'later prompt'),
    ]);

    expect(preview, 'first line');
  });

  test('prompt preview truncates long prompts and returns null without user rows',
      () {
    final long = promptPreview(<TimelineItem>[
      messageWith(id: 'u1', role: MessageRole.user, text: 'x' * 100),
    ]);
    final none = promptPreview(<TimelineItem>[
      messageWith(id: 'a1', role: MessageRole.assistant, text: 'answer'),
    ]);

    expect(long?.length, 60);
    final longValue = long;
    if (longValue != null) {
      expect(longValue.substring(longValue.length - 1), '…');
    }
    expect(none, isNull);
  });
}
